### ISSUE TITLE
CI | Lint | Add `lint-git-history` with options for branch commits structure

### ISSUE THREAD

**Issue by tomerle** [gamitee-dev/public-ci#8]
### Background
1. We are forced to use a merge commit due to release-related automations.
2. Thus we must demand merge commits in GH, and there's no other setting to control it further.
3. We realize that even if it was not for this specific desire, we could use and easy-to-use tool for more control over this.

### Motivation
1. Suggested tool will allow us to easily assert usage of rebase and/or squash (amongst other options).
2. We will implement local git hooks soon, so the annoying/surprising part is only temporary.
3. Seems generic and git-only enough to deserve being cross repository.

### TODO
1. Make it a separate Job so repositories can choose opt-in.
2. Try to be consistent with the function-like naming `lint-git-history`.
3. Suggest `inputs` to enable per-feature opt-in and tuning.
4. A nice general (relevant) `git log ...` output (in case of issue).

**Comment by bentsi** [comment]
1. I also call this "shark teeth" history. From my experience it is not suitable for real life scenarios:
When folks develop a feature and open a feature branch in ideal scenario all the changes are related to that feature, but practically you fix some bugs that you see, improve code and docstrings, all those are not related to the feature itself. Those changes should be tracked in separate commits, so developers can understand easily the motivation behind them from the git history. 
2. I strongly recommend https://pre-commit.com. I can help to integrate it.

**Comment by tomerle** [comment]
### Discussion

_Leaving the implementation of local hooks aside (as we'll coordinate and do it later on)._

_"Not suitable for real life scenarios" is a bit extreme. I can claim the same for any software-related choice. You know... someone once said that C++ is not a worthy programing language. 🙃_ 

Anyway, I'm well aware everything you mentioned, but still... I call YAGNI... 
I don't need anything that full git history can give me, believe we'll never need, and if we will need, then the price for me seems lower than the price of obfuscating the git history.

I also believe encouraging multiple commits encourage a slower old-school approach.
If I'm good enough to create my docstring signature before writing the function body,
Why wait for the function?... Release it!

_I'm calling it "old-school" b/c waiting for the "full feature" resembles Waterfall. Now I know we have manual QA, but we don't do release cycles. We instead try to stick to a continuous workflow that some day will be complimented by true automatic CI/CD._

---

### TODO

Once we will be able to merge PRs w/o a merge commit, there's still the debate of squash vs ff.
I'm a squash kinda guy, as you can understand, but if people pay attention to their commits, I'm all for allowing ff.
But you know... Like you say it should be done... Meaningful commits.
I just don't want to force it, for the reasons above.

So... hearing that... 
Are you ok with me moving on with this, or do you want to discuss it further first?

**Comment by bentsi** [comment]
I don't understand what are you proposing :) 
- Linear history
- Shark Teeth history
?

- regarding "real life scenarios" I am talking from my experience (as Team Leader and developer)
- meaningful commits (title and body) help to understand the reason and motivation for a change, it is enforcable. As you know I come from open source projects and it is absolutely feasable (commitlint is a tool to enforce the commit msg format that uses [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) specification). People come and go, documentation is always neglected, single source of truth is the commit message :)

let's have a meeting on this and discuss

**Comment by tomerle** [comment]
I'll setup a meeting for you and I to discuss, and then we'll share with the team...

**tl;dr**
1. The image on https://www.conventionalcommits.org/en/v1.0.0/ is exactly what I want to avoid.
2. I am looking for qualities. Not a known name. I'm looking for the qualities that are common to linear and shark-teeth.

> Side note:
Our experience matters when no one knows anything about something and we can't afford to learn. Otherwise, it's like saying "putting logic/reasoning aside, I _liked_ it". Which _is_ a lot, no doubt!. But let's agree this is something that experience with is not uncommon/special, so the discussion should be mainly and foremost around login/reasoning.

**Comment by tomerle** [comment]
@bentsi, I'm changing the scope to only add some generic opt-in assertions on the git tree/graph structure. Not forcing anything, and not related to the commit message/code. We _will_ have our meeting, but it's about more than this task is, so I'm separating the concerns in the name of progress. 🤝🚀
